src/ape/managers/plugins.py

```diff
@@ -0,0 +1,178 @@
+import functools
+import importlib
+import pkgutil
+import subprocess
+from typing import Any, Generator, Iterator, List, Optional, Set, Tuple
+
+from ape.__modules__ import __modules__
+from ape.exceptions import ApeAttributeError
+from ape.logging import logger
+from ape.plugins._utils import PIP_COMMAND, clean_plugin_name
+from ape.plugins.pluggy_patch import plugin_manager as pluggy_manager
+from ape.utils.basemodel import _assert_not_ipython_check
+from ape.utils.misc import log_instead_of_fail
+
+
+def valid_impl(api_class: Any) -> bool:
+    """
+    Check if an API class is valid. The class must not have any unimplemented
+    abstract methods.
+
+    Args:
+        api_class (any)
+
+    Returns:
+        bool
+    """
+
+    if isinstance(api_class, tuple):
+        return all(valid_impl(c) for c in api_class)
+
+    # Is not an ABC base class or abstractdataclass
+    if not hasattr(api_class, "__abstractmethods__"):
+        return True  # not an abstract class
+
+    return len(api_class.__abstractmethods__) == 0
+
+
+class PluginManager:
+    _unimplemented_plugins: List[str] = []
+
+    def __init__(self) -> None:
+        self.__registered = False
+
+    @log_instead_of_fail(default="<PluginManager>")
+    def __repr__(self) -> str:
+        return f"<{PluginManager.__name__}>"
+
+    def __getattr__(self, attr_name: str) -> Iterator[Tuple[str, Tuple]]:
+        _assert_not_ipython_check(attr_name)
+
+        # NOTE: The first time this method is called, the actual
+        #  plugin registration occurs. Registration only happens once.
+        self._register_plugins()
+
+        if not hasattr(pluggy_manager.hook, attr_name):
+            raise ApeAttributeError(f"{PluginManager.__name__} has no attribute '{attr_name}'.")
+
+        # Do this to get access to the package name
+        hook_fn = getattr(pluggy_manager.hook, attr_name)
+        hookimpls = hook_fn.get_hookimpls()
+
+        def get_plugin_name_and_hookfn(h):
+            return h.plugin_name, getattr(h.plugin, attr_name)()
+
+        for plugin_name, results in map(get_plugin_name_and_hookfn, hookimpls):
+            # NOTE: Some plugins return a tuple and some return iterators
+            if not isinstance(results, Generator):
+                validated_plugin = self._validate_plugin(plugin_name, results)
+                if validated_plugin:
+                    yield validated_plugin
+            else:
+                # Only if it's an iterator, provider results as a series
+                for result in results:
+                    validated_plugin = self._validate_plugin(plugin_name, result)
+                    if validated_plugin:
+                        yield validated_plugin
+
+    @property
+    def registered_plugins(self) -> Set[str]:
+        self._register_plugins()
+        return {x[0] for x in pluggy_manager.list_name_plugin()}
+
+    @functools.cached_property
+    def _plugin_modules(self) -> Tuple[str, ...]:
```

Comment:
merge miss: https://github.com/ApeWorX/ape/pull/2004

^ missing from the diff. (no longer do we need to parse pip freeze)

- [ ] - i can plop it in